gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/ImmutableFSJobCatalog.java

```diff
@@ -0,0 +1,164 @@
+package gobblin.runtime.job_catalog;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.List;
+import java.util.Properties;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import com.typesafe.config.Config;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.runtime.api.JobCatalog;
+import gobblin.runtime.api.JobCatalogListener;
+import gobblin.runtime.api.JobSpec;
+import gobblin.runtime.api.JobSpecNotFoundException;
+import gobblin.runtime.util.FSJobCatalogHelper;
+import gobblin.util.ConfigUtils;
+import gobblin.util.PathUtils;
+import gobblin.util.PullFileLoader;
+import gobblin.util.filesystem.PathAlterationDetector;
+import gobblin.util.filesystem.PathAlterationObserver;
+
+
+public class ImmutableFSJobCatalog implements JobCatalog {
+
+  protected final FileSystem fs;
+  protected final Config sysConfig;
+
+  protected final JobCatalogListenersList listeners;
+  protected static final Logger LOGGER = LoggerFactory.getLogger(ImmutableFSJobCatalog.class);
+
+  protected final PullFileLoader loader;
+
+  /* The root configuration directory path.*/
+  protected final Path jobConfDirPath;
+  private final FSJobCatalogHelper.JobSpecConverter converter;
+
+  // A monitor for changes to job conf files for general FS
+  // This embedded monitor is monitoring job configuration files instead of JobSpec Object.
+  protected final PathAlterationDetector pathAlterationDetector;
+  protected final PathAlterationObserver observer;
+
+  public ImmutableFSJobCatalog(Config sysConfig)
+      throws Exception {
+    this(sysConfig, null);
```

Comment:
This null issue is addressed by `addPathAlterationObserver` method, which contains `Optional<PathAlterationObserver>` as a parameter. 

The initialization of this parameter, inside the `JobCatalog` constructor is: `Optional<PathAlterationObserver> observerOptional = Optional.fromNullable(this.observer);` 

It could be null here and leave it to the lower layer methods to handle.